File: java/akka-persistence/counter-sample/src/main/java/sample/SampleActor.java
package sample;

import akka.japi.pf.ReceiveBuilder;
import akka.persistence.AbstractPersistentActor;
import akka.persistence.SnapshotOffer;
import scala.PartialFunction;
import scala.runtime.BoxedUnit;

public class SampleActor extends AbstractPersistentActor {
    private final String id;
    private int state = 0;

    public SampleActor(String id) {
        this.id = id;
    }

    @Override
    public String persistenceId() {
        return id;
    }

    @Override
    public PartialFunction<Object, BoxedUnit> receiveRecover() {
        return ReceiveBuilder
                .match(AddedCounter.class, this::updateState)
                .match(SnapshotOffer.class, ss -> state = (Integer)ss.snapshot())
                .build();
    }

    @Override
    public PartialFunction<Object, BoxedUnit> receiveCommand() {
        return ReceiveBuilder
                .match(CounterAdd.class, cmd ->
                        persist(new AddedCounter(cmd.count()), event -> {
                            updateState(event);
                            context().system().eventStream().publish(event);
                        })
                )
                .matchEquals("snapshot", cmd -> saveSnapshot(state))
                .matchEquals("dump", cmd -> System.out.println("counter: " + state))
                .matchEquals("end", cmd -> context().system().terminate())
                .build();
    }

    private void updateState(AddedCounter event) {
        state += event.count();
    }
}

File: kotlin/akka-persistence/counter-sample/src/main/kotlin/sample/SampleActor.kt
package sample

import akka.persistence.AbstractPersistentActor
import akka.persistence.SnapshotOffer

data class CounterAdd(val count: Int)
data class AddedCounter(val count: Int)

class SampleActor(val id: String) : AbstractPersistentActor() {
    private var state = 0

    override fun persistenceId(): String = id

    override fun createReceiveRecover() = receiveBuilder()
            .match(AddedCounter::class.java) { event ->
                updateState(event)
            }
            .match(SnapshotOffer::class.java) { ss ->
                state = ss.snapshot() as Int
            }
            .build()

    override fun createReceive() = receiveBuilder()
            .match(CounterAdd::class.java) { cmd ->
                persist(AddedCounter(cmd.count)) { event ->
                    updateState(event)
                    context.system().eventStream().publish(event)
                }

            }
            .matchEquals("snapshot") {
                saveSnapshot(state)
            }
            .matchEquals("dump") {
                println("counter: ${state}")
            }
            .matchEquals("end") {
                context.stop(self())
            }
            .build()

    private fun updateState(event: AddedCounter) {
        state += event.count
    }
}
